dusk-bytes/src/serialize.rs

```diff
@@ -42,8 +42,56 @@ pub trait DeserializableSlice<const N: usize>: Serializable<N> {
             Self::from_bytes(&bytes)
         }
     }
+
+    /// Deserialize the type reading the bytes from a reader.
+    /// The bytes read are removed from the reader.
+    fn from_reader<R>(buf: &mut R) -> Result<Self, Self::Error>
+    where
+        R: Read,
+        Self: Sized,
+        Self::Error: BadLength,
+    {
+        let mut bytes = [0u8; N];
+        buf.read(&mut bytes)
+            .map_err(|_| Self::Error::bad_length(buf.size(), N))?;
+
+        Self::from_bytes(&bytes)
+    }
 }
 
 // Auto trait [`DeserializableSlice`] for any type that implements
 // [`Serializable`]
 impl<T, const N: usize> DeserializableSlice<N> for T where T: Serializable<N> {}
+
+pub trait Read {
+    fn size(&self) -> usize;
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error>;
+}
+
+impl Read for &[u8] {
+    #[inline]
+    fn size(&self) -> usize {
+        self.len()
+    }
+
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        if buf.len() > self.len() {
+            return Err(Error::bad_length(self.len(), buf.len()));
+        }
+        let amt = buf.len();
+        let (a, b) = self.split_at(amt);
+
+        // First check if the amount of bytes we want to read is small:
+        // `copy_from_slice` will generally expand to a call to `memcpy`, and
+        // for a single byte the overhead is significant.
+        if amt == 1 {
```

Comment:
Its fine here, but in our use-case we will (almost) never fall into serialization / deserialization of single bytes. Hence, we will, in practice, never benefit from this part of the code.

But, as a general purpose crate, it makes sense.

Reply:
We have at least the `NoteType` that is serialized as single byte, and I think we could have similar usage for indicates type / flags as well in the future.

In this way we don't have to make assumption (as we do now), all the types can use `from_reader` (even `NoteType`) - instead that deserialize manually, and the code will be more consistent.